gobblin-metastore/src/main/java/gobblin/metastore/FsStateStore.java

```diff
@@ -73,38 +81,65 @@
-    this.conf = new Configuration();
-    this.fs = FileSystem.get(URI.create(fsUri), this.conf);
-    this.useTmpFileForPut = !FS_SCHEMES_NON_ATOMIC.contains(this.fs.getUri().getScheme());
-    this.storeRootDir = storeRootDir;
-    this.stateClass = stateClass;
+    this(FileSystem.get(URI.create(fsUri), new Configuration()), storeRootDir, stateClass);
+  }
+
+  public FsStateStore(String storeUrl, Class<T> stateClass) throws IOException {
+    this(new Path(storeUrl), stateClass);
   }
 
-  public FsStateStore(FileSystem fs, String storeRootDir, Class<T> stateClass) {
+  public FsStateStore(Path storePath, Class<T> stateClass) throws IOException {
+    this(storePath.getFileSystem(new Configuration()), storePath.toUri().getPath(), stateClass);
+  }
+
+  public FsStateStore(FileSystem fs, String storeRootDir, Class<T> stateClass) throws IOException {
     this.fs = fs;
     this.useTmpFileForPut = !FS_SCHEMES_NON_ATOMIC.contains(this.fs.getUri().getScheme());
     this.conf = this.fs.getConf();
     this.storeRootDir = storeRootDir;
     this.stateClass = stateClass;
   }
 
-  public FsStateStore(String storeUrl, Class<T> stateClass) throws IOException {
-    this.conf = new Configuration();
-    Path storePath = new Path(storeUrl);
-    this.fs = storePath.getFileSystem(this.conf);
-    this.useTmpFileForPut = !FS_SCHEMES_NON_ATOMIC.contains(this.fs.getUri().getScheme());
-    this.storeRootDir = storePath.toUri().getPath();
-    this.stateClass = stateClass;
-  }
-
+  /**
+   * Create a new store.
+   *
+   * <p>
+   *     A store that does not exist will be created when any put
+   *     method is called against it.
+   * </p>
+   *
+   * @param storeName store name
+   * @return if the store is successfully created
+   * @throws IOException
+   */
   @Override
   public boolean create(String storeName) throws IOException {
+    Preconditions.checkArgument(!Strings.isNullOrEmpty(storeName), "Store name is null or empty.");
+
     Path storePath = new Path(this.storeRootDir, storeName);
     return this.fs.exists(storePath) || this.fs.mkdirs(storePath, new FsPermission((short) 0755));
   }
 
+  /**
+   * Create a new table in a store.
+   *
+   * <p>
+   *     A table that does not exist will be created when any put
+   *     method is called against it.
+   * </p>
+   *
+   * @param storeName store name
+   * @param tableName table name
+   * @return if the table is successfully created
+   * @throws IOException
+   */
   @Override
   public boolean create(String storeName, String tableName) throws IOException {
+    Preconditions.checkArgument(!Strings.isNullOrEmpty(storeName), "Store name is null or empty.");
+    Preconditions.checkArgument(!Strings.isNullOrEmpty(tableName), "Table name is null or empty.");
+    Preconditions.checkArgument(!isCurrentFile(tableName),
+        String.format("Table name is %s", StateStoreTableInfo.CURRENT_NAME));
```

Comment:
The problem is that without putting an understanding of "CURRENT" in the state store, all calling code will have to change.  This way, any calling code can as for the file named "CURRENT" and they get the latest one.